### PR TITLE
DEV-32951 - add FieldsNamesSignatureProcessor

### PR DESCRIPTION
**Pull request by ronshay**

### CODE REVIEW

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@ProcessorProvider(type = "docSignature", factory = DocSignatureProcessor.Factory.class)
+public class DocSignatureProcessor implements Processor {
+    private final SignatureMode signatureMode;
+    private final Set<String> includeValueFields;
+    private final String DOC_SIGNATURE_FIELD = "logzio_doc_signature";
```
Comment (ronshay) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:20]: move field name to config

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        try {
+            signature = createSignature(doc);
+        } catch (Exception e) {
+            return ProcessResult.failure(
+                    "failed to create signature, SignatureMode: " + signatureMode,
+                    new ProcessorExecutionException(DocSignatureProcessor.class.getSimpleName(), e));
+        }
+
+        if(signature.isEmpty())
+            return ProcessResult.failure("signature is empty, SignatureMode: " + signatureMode);
```
Comment (ronshay) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:38]: consider

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        @Override
+        public DocSignatureProcessor create(Map<String,Object> config) {
+            DocSignatureProcessor.Configuration fieldsNamesSignatureConfig =
+                    JsonUtils.fromJsonMap(DocSignatureProcessor.Configuration.class, config);
+            return new DocSignatureProcessor(fieldsNamesSignatureConfig.getSignatureMode(),
+                    fieldsNamesSignatureConfig.getIncludeValueFields());
+        }
+    }
+
+    public static class Configuration implements Processor.Configuration {
```
Comment (ronshay) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:118]: add validation

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private String getFieldsValuesString(Doc doc) {
+        List<String> values = includeValueFields.stream().collect(Collectors.toList())
+                .stream()
+                .filter(doc::hasField)
+                .map(doc::getField)
+                .map(Object::toString)
+                .collect(Collectors.toList());
+        return values.isEmpty() ? "" : JsonUtils.toJsonString(values);
```
Comment (ronshay) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:72]: redundant json

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private Set<String> extractFieldsNames(Doc doc) throws InterruptedException {
```
Comment (ronshay) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:75]: rename

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        Map<String, Object> source = doc.getSource();
+        Set<String> fields = new HashSet<>();
+        extractFieldsNames(source, null, fields);
+        return fields;
+    }
+
+    public void extractFieldsNames(Object object, String key, Set<String> fields) throws InterruptedException {
```
Comment (AvishayHirsh) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:82]: rename to parent

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        if (Thread.currentThread().isInterrupted()) throw new InterruptedException();
+        if(object instanceof Map) {
+            Map<String, Object> map = (Map) object;
+            for(Map.Entry<String, Object> entry : map.entrySet()) {
+                extractFieldsNames(entry.getValue(),
+                        key != null ?
+                                new StringBuilder(key).append('.').append(entry.getKey()).toString() : entry.getKey(),
```
Comment (AvishayHirsh) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:89]: extract this line to separated line with meaningful var name

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    private String getFieldsNamesString(Doc doc) throws InterruptedException {
+        Set<String> fieldsNames = extractFieldsNames(doc);
+        return fieldsNames.isEmpty() ? "" : JsonUtils.toJsonString(fieldsNames);
+    }
+
+    private String getFieldsValuesString(Doc doc) {
+        List<String> values = includeValueFields.stream().collect(Collectors.toList())
+                .stream()
+                .filter(doc::hasField)
```
Comment (AvishayHirsh) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:68]: add test that nested field names is supported

**Review by AvishayHirsh — CHANGES_REQUESTED** [review]

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java (hunk 2 above)
Comment (ronshay) [reply sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:38]: i think we should keep it.. consider a scenario where the algorithm is broken but does not throw an exception...

**Comment by AvishayHirsh** [comment]
Add documentation for this processor

File: sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            case FIELDS_NAMES:
+                return new ArrayList<>(extractFieldsNames(doc));
+            case HYBRID:
+                return Stream.concat(getFieldsValues(doc).stream(), extractFieldsNames(doc)
+                        .stream()).collect(Collectors.toList());
+            default: return Collections.emptyList();
+        }
+    }
+
+    private void addSignatureField(Doc doc, List<String> signatureCollection) {
```
Comment (AvishayHirsh) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/DocSignatureProcessor.java:64]: 1] so we agree that it is better to use set hashcode 
2] we want to add a test that {a,b,c} signature is equals to {c,b,a}

File: sawmill-core/src/main/java/io/logz/sawmill/processors/AddSignatureProcessor.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    private Set<String> extractFieldsNames(Doc doc) throws InterruptedException {
+        Map<String, Object> source = doc.getSource();
+        Set<String> fields = new HashSet<>();
+        extractFieldsNames(source, null, fields);
+        return fields;
+    }
+
+    public void extractFieldsNames(Object object, String parentKey, Set<String> fields) throws InterruptedException {
```
Comment (AvishayHirsh) [inline sawmill-core/src/main/java/io/logz/sawmill/processors/AddSignatureProcessor.java:87]: change to private

**Review by AvishayHirsh — APPROVED** [review]
Great work

**Pull request merged**